File: vue-vben-admin/apps/web-antd/src/api/core/menu.ts
import type { RouteRecordStringComponent } from '@vben/types';

import { requestClient } from '#/api/request';

interface BackendMenu {
  id: string;
  parentId: string;
  title: string;
  type: number;
  path: string;
  component: string;
  permission: string;
  icon: string;
  sort: number;
  isShow: number;
  isCache: number;
  linkUrl: string;
  status: number;
  children?: BackendMenu[];
}

/** 将后端菜单转换为 Vben 路由格式 */
function transformMenus(menus: BackendMenu[]): RouteRecordStringComponent[] {
  return menus
    .filter((m) => m.isShow === 1)
    .map((menu) => {
      const route: RouteRecordStringComponent = {
        name: menu.path?.replace(/\//g, '-').replace(/^-/, '') || `menu-${menu.id}`,
        path: menu.path || '',
        component: menu.component || '',
        meta: {
          title: menu.title,
          icon: menu.icon || undefined,
          order: menu.sort,
          hideInMenu: menu.isShow !== 1,
          keepAlive: menu.isCache === 1,
          authority: menu.permission ? [menu.permission] : undefined,
        },
      };

      if (menu.children?.length) {
        route.children = transformMenus(menu.children);
      }

      // 外链
      if (menu.type === 4 && menu.linkUrl) {
        route.meta = { ...route.meta, link: menu.linkUrl };
      }

      return route;
    });
}

/**
 * 获取用户所有菜单
 */
export async function getAllMenusApi() {
  const res = await requestClient.get<{ menus: BackendMenu[] }>(
    '/system/auth/menus',
  );
  const menus = res?.menus ?? [];
  return transformMenus(menus);
}

File: app/system/internal/model/menu.go
package model

import (
	"gbaseadmin/utility/snowflake"
)

// Menu DTO 模型

// MenuCreateInput 创建菜单表输入
type MenuCreateInput struct {
	ParentID snowflake.JsonInt64 `json:"parentID"`
	Title string `json:"title"`
	Type int `json:"type"`
	Path string `json:"path"`
	Component string `json:"component"`
	Permission string `json:"permission"`
	Icon string `json:"icon"`
	Sort int `json:"sort"`
	IsShow int `json:"isShow"`
	IsCache int `json:"isCache"`
	LinkURL string `json:"linkURL"`
	Status int `json:"status"`
}

// MenuUpdateInput 更新菜单表输入
type MenuUpdateInput struct {
	Id snowflake.JsonInt64 `json:"id"`
	ParentID snowflake.JsonInt64 `json:"parentID"`
	Title string `json:"title"`
	Type int `json:"type"`
	Path string `json:"path"`
	Component string `json:"component"`
	Permission string `json:"permission"`
	Icon string `json:"icon"`
	Sort int `json:"sort"`
	IsShow int `json:"isShow"`
	IsCache int `json:"isCache"`
	LinkURL string `json:"linkURL"`
	Status int `json:"status"`
}

// MenuDetailOutput 菜单表详情输出
type MenuDetailOutput struct {
	ParentID snowflake.JsonInt64 `json:"parentID"`
	Title string `json:"title"`
	Type int `json:"type"`
	Path string `json:"path"`
	Component string `json:"component"`
	Permission string `json:"permission"`
	Icon string `json:"icon"`
	Sort int `json:"sort"`
	IsShow int `json:"isShow"`
	IsCache int `json:"isCache"`
	LinkURL string `json:"linkURL"`
	Status int `json:"status"`
	CreatedAt *gtime.Time `json:"createdAt"`
	UpdatedAt *gtime.Time `json:"updatedAt"`
}

// MenuListOutput 菜单表列表输出
type MenuListOutput struct {
	ParentID snowflake.JsonInt64 `json:"parentID"`
	Title string `json:"title"`
	Type int `json:"type"`
	Path string `json:"path"`
	Component string `json:"component"`
	Permission string `json:"permission"`
	Icon string `json:"icon"`
	Sort int `json:"sort"`
	IsShow int `json:"isShow"`
	IsCache int `json:"isCache"`
	LinkURL string `json:"linkURL"`
	Status int `json:"status"`
	CreatedAt *gtime.Time `json:"createdAt"`
	UpdatedAt *gtime.Time `json:"updatedAt"`
}

// MenuListInput 菜单表列表查询输入
type MenuListInput struct {
	PageNum  int `json:"pageNum"`
	PageSize int `json:"pageSize"`
}

// MenuTreeOutput 菜单表树形输出
type MenuTreeOutput struct {
	ParentID snowflake.JsonInt64 `json:"parentID"`
	Title string `json:"title"`
	Type int `json:"type"`
	Path string `json:"path"`
	Component string `json:"component"`
	Permission string `json:"permission"`
	Icon string `json:"icon"`
	Sort int `json:"sort"`
	IsShow int `json:"isShow"`
	IsCache int `json:"isCache"`
	LinkURL string `json:"linkURL"`
	Status int `json:"status"`
	Children []*MenuTreeOutput `json:"children"`
}

